cmd/logmonitor/db.go

```diff
@@ -0,0 +1,147 @@
+)
+
+// initialEmptyRoot represents an empty trusted root and is used when no
+// information is known about the log being monitored.
+var initialEmptyRoot TrustedLogRoot
+
+func init() {
+	// Set up initialEmptyRoot, which will get used when the database does not
+	// have an existing trusted log root (ie. the first time a tree is
+	// monitored).
+	emptyRoot := &types.LogRootV1{
+		TreeSize: 0,
+	}
+	marshalled, err := emptyRoot.MarshalBinary()
+	if err != nil {
+		glog.Exitf("Could not marshal the empty root: %v\n", err)
+	}
+	initialEmptyRoot = TrustedLogRoot{MarshalledRoot: marshalled, TreeSize: 0}
+}
+
+// TrustedLogRoot is the database model used for storing information about
+// trusted log roots.
+type TrustedLogRoot struct {
+	LogID          int64 `gorm:"primary_key"`
+	MarshalledRoot []byte
+	TreeSize       uint64
+
+	CreatedAt time.Time
+	UpdatedAt time.Time
+}
+
+// Open opens a new database connection
+func Open(database, databaseURI string) (*gorm.DB, error) {
+	uri, err := preprocessURIFor(database, databaseURI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to process the database URI: %v", err)
+	}
+
+	db, err := gorm.Open(database, uri)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open database: %v", err)
+	}
+
+	return db, nil
+}
+
+// preprocessURIFor does any necessary changes to the connection URI in order
+// to make sure the database connection behaves as expected.
+func preprocessURIFor(database, databaseURI string) (string, error) {
+	u, err := url.Parse(databaseURI)
+	if err != nil {
+		return "", fmt.Errorf("could not parse the database URI: %v", err)
+	}
+
+	if database == "mysql" {
+		// Make the go mysql driver's output type used for DATE and DATETIME values
+		// be time.Time instead of []byte.
+		q := u.Query()
+		q.Set("parseTime", "true")
+		u.RawQuery = q.Encode()
+	}
+
+	return u.String(), nil
+}
+
+// Setup ensures that the database tables are set up properly.
+func Setup(db *gorm.DB) error {
+	if err := db.AutoMigrate(&TrustedLogRoot{}).Error; err != nil {
+		return fmt.Errorf("failed to migrate the trusted_log_roots table: %v", err)
+	}
+
+	return nil
+}
+
+// GetCurrentRootFor returns the current trusted root from the database.
+func GetCurrentRootFor(db *gorm.DB, logID int64) (*types.LogRootV1, error) {
+	trustedLogRoot := TrustedLogRoot{}
+
+	query := db.Where(TrustedLogRoot{LogID: logID})
+	query = query.Attrs(initialEmptyRoot) // Values to write if the record is not found.
+	if err := query.FirstOrCreate(&trustedLogRoot).Error; err != nil {
+		return nil, fmt.Errorf("failed to get or create the current trusted root for log %d: %v", logID, err)
+	}
+
+	if trustedLogRoot.TreeSize == 0 {
+		glog.Warningf("Using an empty root for log %d", logID)
+	} else {
+		glog.Infof("Loaded a root for log %d with tree size = %d", logID, trustedLogRoot.TreeSize)
+	}
+
+	currentRoot := &types.LogRootV1{}
+	if err := currentRoot.UnmarshalBinary(trustedLogRoot.MarshalledRoot); err != nil {
```

Comment:
`UnmarshalBinary` is considered a private method.  Please migrate to `client.LogVerifier.VerifyRoot`